_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+One of the key features of Kafka is the support for transactions, which provides exactly-once semantics (EOS) and is available since Kafka 0.11 ([KIP-98](https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging)).
+The support for EOS was recently extended to source connectors in Kafka Connect, which enables configuration-driven and fully transactional streaming pipelines ([KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors)).
+
+As messages flow through Kafka, they can be processed by multiple consumers, and ensuring that all consumers that should receive a message actually receive it once and only once can be crucial.
+The EOS guarantees that any set of operations inside Kafka will be completed atomically, either all of the operations succeed or none of them do.
+This is done by defining transaction boundaries that dictate which operations can be grouped together in a single transaction.
+These operations can include reading from topics, writing to topics, and committing offsets.
+A failure during processing can cause a transaction to be aborted, in which case none of the messages from the transaction will be readable by consumers.
+This helps in maintaining high levels of reliability and consistency as data is processed throughout the application's lifetime.
+
+<!--more-->
+
+### Delivery guarantees
+
+In distributed systems like Kafka, any component can fail independently and you have to deal with this fact.
+A broker may crash or a network failure may happen while the producer is sending a message to a topic.
+
+Depending on the configuration, you can have three different delivery semantics:
+
+- **at-most-once**: the producer does not retry when there is a timeout or error (data loss risk)
+- **at-least-once**: the producer retries but the broker can fail after writing but before sending the ack (duplicates risk)
+- **exactly-once**: even if the producer retries sending a message, it is written exactly once (idempotency)
+
+With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
+There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. financial institutions).
+
+To be fair, there is no such thing as exactly-once delivery in distributed systems (two generals problem), but we can fake it by implementing idempotent operations.
+Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.
+
+Since Kafka 3.0, the producer enables the strongest delivery guarantees by default (`acks=all`, `enable.idempotence=true`).
```

Comment:
I think the transactional producer gives the strongest semantics. With idempotency these guarantees are only valid per message and for the lifetime of the producer instance.